lib/core/src/display/rendering/transform.rs

```diff
@@ -0,0 +1,172 @@
+    pub scale       :       Vector3<f32>,
+}
+
+impl Transform {
+    /// Creates an identity transform.
+    pub fn identity() -> Self {
+        Self {
+            translation : Vector3::new(0.0, 0.0, 0.0),
+            rotation :    UnitQuaternion::identity(),
+            scale :       Vector3::new(1.0, 1.0, 1.0),
+        }
+    }
+
+    /// Sets Transform's translation.
+    pub fn set_translation(&mut self, x: f32, y: f32, z: f32) {
+        self.translation = Vector3::new(x, y, z);
+    }
+
+    /// Set Transform's scale.
+    pub fn set_scale(&mut self, x: f32, y: f32, z: f32) {
+        self.scale = Vector3::new(x, y, z);
+    }
+
+    /// Set Transform's rotation from Euler angles in radians.
+    pub fn set_rotation(&mut self, roll: f32, pitch: f32, yaw: f32) {
+        self.rotation = from_euler_angles_pry(roll, pitch, yaw);
+    }
+
+    /// Gets a homogeneous transform Matrix4. The rotation order is YXZ (pitch,
+    /// roll, yaw).
+    // Note [Transform to Matrix4 composition]
+    // =======================================
+    // based on
+    // https://github.com/mrdoob/three.js/blob/master/src/math/Matrix4.js#L732
+    pub fn to_homogeneous(&self) -> Matrix4<f32> {
+        let (x, y, z, w) = (
+            self.rotation.coords.x,
+            self.rotation.coords.y,
+            self.rotation.coords.z,
+            self.rotation.coords.w,
+        );
+        let (x2, y2, z2) = (x + x, y + y, z + z);
+        let (xx, xy, xz) = (x * x2, x * y2, x * z2);
+        let (yy, yz, zz) = (y * y2, y * z2, z * z2);
+        let (wx, wy, wz) = (w * x2, w * y2, w * z2);
+
+        let (sx, sy, sz) = (self.scale.x, self.scale.y, self.scale.z);
+
+        let m00 = (1.0 - (yy + zz)) * sx;
+        let m01 = (xy + wz) * sx;
+        let m02 = (xz - wy) * sx;
+        let m03 = 0.0;
+
+        let m10 = (xy - wz) * sy;
+        let m11 = (1.0 - (xx + zz)) * sy;
+        let m12 = (yz + wx) * sy;
+        let m13 = 0.0;
+
+        let m20 = (xz + wy) * sz;
+        let m21 = (yz - wx) * sz;
+        let m22 = (1.0 - (xx + yy)) * sz;
+        let m23 = 0.0;
+
+        let (m30, m31, m32) = (self.translation.x,
+                               self.translation.y,
+                               self.translation.z);
+        let m33 = 1.0;
+        Matrix4::new(m00, m10, m20, m30,
+                     m01, m11, m21, m31,
+                     m02, m12, m22, m32,
+                     m03, m13, m23, m33)
+    }
+}
+
+// ======================
+// === Matrix Printer ===
+// ======================
+pub trait IntoCSSMatrix {
+    fn into_css_matrix(&self) -> String;
+}
+
+impl IntoCSSMatrix for Matrix4<f32> {
+    fn into_css_matrix(&self) -> String {
+        let mut iter = self.iter();
+        let item = iter.next().expect("Matrix4 should have the first item");
+        let acc = format!("matrix3d({}", item);
+        let mut ret = iter.fold(acc, |acc, item| format!("{}, {}", acc, item));
+        ret.push(')');
```

Comment:
This is ugly. `push(')')` is hacky. First concatenate children, then do `format!("matrix3d({})", ...)` instead.